### ISSUE TITLE
Adding GAN support to TensorGraph

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#535]
I think our current model with `TensorGraph` can't nicely support GANs. In order to train GANs, you need to train both a discriminator `D` and a generator `G`. The training of these two models is threaded together, with each updated for a few minibatches at a time. While updating the generator, the discriminator weights should be frozen. While updating the discriminator, the generator weights need to be frozen.

I have an idea for a new `TensorGraph` API to support this sort of training. We add a new `layers` kwarg to `fit`:

- `model.fit(feed_generator, layers=None)`

The idea is that `fit` can be instructed to train only some layers at a time. Note that there's now a semantic change:

- `Layer` is responsible for tracking the variables it introduces. 
  - `layer.variables` should be a list of all trainable variables created by that `Layer`.

@peastman and @lilleswing Does this basic design make sense? 

Hat tip to @enf for explaining GAN training to me in detail!

**Comment by pierrebaque** [comment]
Hello,

I was wondering if you had a specific idea about what GANs could be useful for, it terms of practical applications in DeepChem?

Thanks

**Comment by rbharath** [comment]
It might be useful to implement interesting models such as OR-GAN https://arxiv.org/abs/1705.10843

**Comment by peastman** [comment]
I'm starting to look into how we can better support GANs.  The API suggested above sounds reasonable as far as it goes, but I'm not sure it's sufficient.  For example, don't you often have different loss functions for training the generator and discriminator?  Also, "GANs" are really a broad category of many different sorts of models.  There's so many variations on the basic idea: wGAN, cGAN, CycleGAN, etc.  So we should consider two distinct questions:

1. What core infrastructure do we need to make it *possible* to support any of them?
2. Are there commonalities between them, such that we could write one or more convenience classes that make it *easy* to implement the most important types?

**Comment by rbharath** [comment]
I think the general infrastructure is the ability to train "submodels" (or "metalayers") in the same graph with different loss functions.

**Comment by peastman** [comment]
Let's think about what that means.

- Submodels are non-overlapping subsets of the layers in a model.  (Is that reasonable?  Would we ever want to treat a single layer as being part of two different submodels?)
- Each submodel has its own loss function.
- Each submodel has its own optimizer.
- All submodels are part of the same `tf.Graph`, and their variables are stored in the same `tf.Session`.
- Do submodels always get optimized separately?  Or would we want the ability to optimize multiple submodels together as a unit (so that a single optimization step updates all of them)?
- How do submodels relate to inputs and outputs?  When you call `fit()`, you provide the `X` and `y` values.  Are those values specific to a particular submodel, such that you might use different numbers and shapes of inputs when training different submodels?  Or do they apply to the model as a whole?

**Comment by rbharath** [comment]
I think submodels do overlap in GANs, so we should allow for overlapping layers. I think it's safe to assume that only one submodel is optimized per update (not entirely sure on this one though).

**Comment by peastman** [comment]
How about this as a candidate API:

```python
submodel = model.create_submodel(layers, loss, optimizer)
model.fit(dataset, submodel=submodel)
```

So a submodel specifies a list of layers, a loss function, and an optimizer.  When you perform fitting, you can optionally specify a submodel to do the fitting on.  It then uses a different `train_op`, but everything else works exactly the same.

**Comment by rbharath** [comment]
This seems like a solid API to get us started!

**Comment by peastman** [comment]
We'll need a few other changes too.  As noted in #789, we need to change how the queue is synchronized.  Right now, you can only ever call `fit()` once on a model.  If you call it a second time, it throws an exception.  Also, suppose you tried to call `fit_batch()` repeatedly, alternating between the generator and discriminator.  That would cause it to save a checkpoint and print out the current loss after every batch, which would be crazy slow.  We need a way to restrict that.  A simple solution would be a flag to tell it not to print output or write checkpoints.  But then it wouldn't write any checkpoints at all, so you would need to remember to periodically omit that flag, or perhaps have a separate method to explicitly write a checkpoint?  Also, it would no longer be possible to collect averaged statistics on the loss, since every call to `fit()` would only involve a single batch.  Maybe have `fit()` return the current loss so you can average it yourself?

Alternatively we could do something much fancier where you define "fitting protocols" within the TensorGraph.  But it's probably simpler to just provide the raw materials, and let other classes use them to define whatever protocols they want.

**Comment by peastman** [comment]
@joegomes I'd especially value your feedback on this, since you've been working with GANs recently.

**Comment by joegomes** [comment]
One trick that could be useful is [tf.control_dependencies](https://www.tensorflow.org/api_docs/python/tf/control_dependencies).  The idea would be to define the discrim_train_op and use it as the control input, then, within tf.control_dependencies([discrim_train_op]), define the gen_train_op.  You can then fetch both discrim_train_op and gen_train_op within the same sess.run call and they will evaluate in succession.  This could help keep from calling `fit_batch()` repeatedly.

**Comment by peastman** [comment]
What would that look like in terms of the TensorGraph API?  Are you suggesting some sort of "fitting protocols" like I mentioned above?  Is the goal to improve performance, or is there something else you're trying to address?

**Comment by joegomes** [comment]
Yeah, this would fall under fitting protocol and the goal would be to improve performance.

**Comment by peastman** [comment]
Is there any advantage over just calling `session.run([discriminator_train_op, generator_train_op])`?  We could do that by allowing the `submodel` argument to take a list of submodels.

As far as performance is concerned, I think there are two main concerns:

1. If we evaluate both training ops in a single call to `run()`, the feed dict values only have to be transferred to the GPU once.  If we evaluate them separately, the values will get transferred twice.  Note though that training the generator involves very little data, so this may be a minor consideration.

2. If we only call `fit_batch()` to fit on a single batch at a time, it can't make use of the queue.  If we let TensorGraph handle iterating through batches, it can use the queue to reduce latency.

That said, I've been looking through various GAN implementations in Tensorflow, and every one I've found so far just alternates calling `run()` once to train the discriminator and once to train the generator.

I'm wondering if "submodel" is really the right word.  Perhaps "training objective" would be more accurate?  I'm not sure which would be less confusing.

**Comment by rbharath** [comment]
I like "submodel" better since it emphasizes that different parts of the model are being trained. Not a hard preference though.

**Comment by peastman** [comment]
I'm not sure how best to implement a GAN with TensorGraph.  What is the input to the discriminator?  I need to be able to train it with two different types of batches: ones that were generated by the generator, and ones that came from the training data?

I could, of course, implement them as two completely separate models.  Make the input to the discriminator be a Feature layer, which I can supply with either real data or generated data.  But that's going to be inefficient.  It will be much more efficient if the generated data can be routed straight from the generator to the discriminator without a round trip through Python.  But then how do I train it on real data?

Or I could make two separate copies of the discriminator with shared variables between them.  That could work, but it's a bit ugly, and currently Dense is the only layer that supports shared variables.

In principle I should be able to route the generator straight into the discriminator, then use the feed dict to replace it with a constant when training on real data.  But I'm not sure that will actually work.  In the past, I've had issues using a feed dict to replace tensors that weren't placeholders.  The documentation claims that it should work, but my experience doesn't agree.

**Comment by peastman** [comment]
Never mind, I think I see how it has to work.  Each batch should include both real and generated samples, so I need to concatenate the two tensors and send them into the discriminator together.

**Comment by peastman** [comment]
I think this is going to be ugly no matter how I do it.  There are several different losses that need to be computed: the discriminator loss on the real data, the discriminator loss on the generated data, and the generator loss on the generated data.  And I need to compute all of those based on a single copy of the discriminator with a single set of inputs.

I've looked at various GAN implementations, and often they do have multiple copies of the discriminator.  It's made me recognize a difference between Tensorflow and TensorGraph that I don't think I really appreciated before.  In Tensorflow you define variables, and then you do calculations based on them.  Those are two separate steps.  You can wrap up the calculations in a function, so repeating the same calculations with the same variables but different inputs just means calling the function twice.

TensorGraph is different.  It combines variables and calculations into a single "layer" object.  That makes the easy cases easier, but I think it may make the complicated cases harder.

**Comment by rbharath** [comment]
I think we're running into the same issues that the Keras team has run into that some of these things are tricky to implement while keeping to layer abstractions. We might want to follow along the implementation style from https://github.com/roatienza/Deep-Learning-Experiments/blob/master/Experiments/Tensorflow/GAN/dcgan_mnist.py which looks relatively clean.

This discussion also makes me think about this medium post: https://medium.com/@devnag/generative-adversarial-networks-gans-in-50-lines-of-code-pytorch-e81b79659e3f. I think complicated models are just easier in PyTorch sometimes and we might run into more cases like this one as we start building increasingly sophisticated models.

**Comment by peastman** [comment]
> We might want to follow along the implementation style from

It looks like they're depending on a rather different abstraction than anything TensorGraph provides.  The Sequential model is a sequence of layers, where the output of one layer is the input to the next.  A layer doesn't know what its input is.  You can include a layer in multiple models, and it may have a different input in each one.

> I think complicated models are just easier in PyTorch sometimes

I think you're looking for any excuse to push PyTorch, even when it's irrelevant to the question. :)  The issue here is TensorGraph, not Tensorflow.  It's also easy to implement GANs with Tensorflow.  The problem is in our own code, which only supports a more limited set of abstractions.

**Comment by rbharath** [comment]
Should we perhaps add on a sequential model abstraction on top of TensorGraph?

Guilty as charged on PyTorch :). Read my case in the other thread #843.

**Comment by peastman** [comment]
I've created an example that trains a CGAN.  Take a look and see what you think:

https://github.com/peastman/deepchem/blob/0ad056ff123fad54be6aac9d82398056b572c900/examples/notebooks/Conditional_GAN.ipynb

The code isn't too bad, but I also think I could make a specialized GAN class that would simplify it a lot more.  You'd still be able to implement everything by hand as shown here, but for common cases it would take care of a lot of the details for you, like defining the loss functions and running the training loop.

How general should that GAN class be?  Or rather, which [GAN variants](https://deephunt.in/the-gan-zoo-79597dc8c347) should it try to support?

**Comment by rbharath** [comment]
@peastman [DCGAN](https://github.com/carpedm20/DCGAN-tensorflow) and [ORGAN](https://arxiv.org/abs/1705.10843) would be great classes to have to start.

**Comment by peastman** [comment]
DCGAN is easy, since it's a completely ordinary GAN.  It just specifies details of how to design the generator and discriminator.

ORGAN might be beyond the scope of this.  It combines the GAN with RL, which is complicated in itself.  I'm not sure if I can support that in a generic GAN class.

With the design I'm thinking of right now, here are the things you could customize:

1. The inputs (noise input to the generator, training data to the discriminator, conditional inputs to both).
2. The generator and discriminator, defined by TensorGraph layers.
3. The loss functions.  It will define standard GAN loss functions for you, but you can add extra terms or replace them altogether.

**Comment by rbharath** [comment]
DCGAN would be a great place to start.

We might want to plan for ORGAN support as a different standalone class too since many of our users will be very interested in using GANs for molecular design purposes, but that sounds like we should discuss on a separate thread.

**Comment by peastman** [comment]
Ok, sounds good.  Do we have any datasets that are appropriate for applying DCGAN to?  If so, that would make a good example.

**Comment by pierrebaque** [comment]
There might be a slight technical issue in applying DCGAN directly. Here, the nature of the object we want to generate is inherently a discrete object, whereas images are continuous. I think that was one of the main motivation behind using RL for ORGAN.
That being said, I have experiments where I can learn generate and solve Sudoku with GANs, which is a discrete problem as well. For this, I am using Wasserstein-GANs with Gradient Penalty, which might be the best choice here. 
What do you think?

**Comment by peastman** [comment]
Depends on what your application is.  There are plenty of chemistry problems that involve continuous data (electron densities, spectra, etc.) for which DCGAN could make sense.

WGAN is another good one to try to support.  I think that should work with the system I outlined above.  Do you see any issues with it?  From what I can tell, it works just like a standard GAN but with a different loss function.  What is the gradient penalty you mentioned?

**Comment by pierrebaque** [comment]
Hello,
Thanks for your reply, I hadn't thought of the applications you mention.

I don't see any issue in using WGAN in your system. I see it as a way to constrain the discriminator so that it cannot focus too fast on small details to take the best over the generator. 
The gradient penalty is the one which is described in this paper [Improved Training of Wasserstein GANs](https://arxiv.org/abs/1704.00028), which works very well in practice. It is quite easy to implement using tensorflow.

**Comment by peastman** [comment]
Thanks, I'll take a look at that paper.

**Comment by rbharath** [comment]
@peastman I think it might be easiest to just start with the faces example dataset from https://github.com/carpedm20/DCGAN-tensorflow. We can probably figure out chemically relevant datasets after the core implementation is debugged.

**Comment by peastman** [comment]
I'm now looking into what would be involved in supporting ORGAN.  What aspects of it are you interested in?  This passage from the paper summarizes what it's about:

> "Note that our implementation on top of SeqGAN is merely because we are working with sequential data. In theory, the ORGAN model can be used with most types of GAN."

Really ORGAN is nothing more than adding another term to the loss function.  They use RL because they're generating sequence data with SeqGAN, and SeqGAN is based on RL.  What exactly did you mean when you suggested supporting ORGAN?  Did you mean the ability to add an extra term to the loss (very easy), or did you really mean SeqGAN for generating sequence data?

**Comment by rbharath** [comment]
Using SeqGAN to generate sequence data is probably the right underlying primitive. I imagine that users would be interested in generating new SMILES strings for new molecules to synthesize or test and having SeqGAN support would make their lives easier.

**Comment by pierrebaque** [comment]
In the abstract of SeqGAN, they also say the following, which could be very important in my view:

> " A major reason lies in that the discrete outputs from
> the generative model make it difficult to pass the gradient update
> from the discriminative model to the generative model."

I think that the model that should be used really depends on the kind of data we want to generate. Maybe would it be better to focus on a specific chemical application first?
